models/lorenz_96_tracer_advection/model_mod.f90

```diff
@@ -0,0 +1,533 @@
+
+end subroutine initialize
+
+!------------------------------------------------------------------
+!> Perturbs a model state for generating initial ensembles.
+!> Returning interf_provided .true. means this code has
+!> added uniform small independent perturbations to a
+!> single ensemble member to generate the full ensemble.
+subroutine pert_model_copies(state_ens_handle, ens_size, pert_amp, interf_provided)
+
+type(ensemble_type), intent(inout) :: state_ens_handle
+integer,   intent(in) :: ens_size
+real(r8),  intent(in) :: pert_amp
+logical,  intent(out) :: interf_provided
+
+integer :: i,j, num_my_grid_points
+integer(i8), allocatable :: my_grid_points(:)
+type(location_type) :: location
+integer :: var_type
+
+interf_provided = .true.
+
+call init_random_seq(random_seq, my_task_id()+1)
+! if we are running with more than 1 task, then
+! we have all the ensemble members for a subset of
+! the model state.  which variables we have are determined
+! by looking at the global index number into the state vector.
+
+! how many grid points does my task have to work on?
+! and what are their indices into the full state vector?
+num_my_grid_points = get_my_num_vars(state_ens_handle)
+allocate(my_grid_points(num_my_grid_points))
+call get_my_vars(state_ens_handle, my_grid_points)
+
+do i=1,num_my_grid_points
+    call get_state_meta_data(my_grid_points(i), location, var_type)
+
+    if (var_type == QTY_STATE_VARIABLE) then
+        do j=1,ens_size
+            state_ens_handle%copies(j, i) = random_gaussian(random_seq, state_ens_handle%copies(j, i), pert_amp)
+        end do
+    !Perturbing all source grid points
+    else if (var_type == QTY_TRACER_SOURCE) then
+        do j=1,ens_size
+         state_ens_handle%copies(j, i) = state_ens_handle%copies(j, i) + random_gaussian(random_seq, 0.00_r8, 50.00_r8)
+        end do
+    end if
+end do
+
+deallocate(my_grid_points)
+end subroutine pert_model_copies
+
+!------------------------------------------------------------------
+! Writes model-specific attributes to a netCDF file
+
+subroutine nc_write_model_atts(ncid, domain_id)
+
+integer, intent(in) :: ncid
+integer, intent(in) :: domain_id
+
+integer :: msize
+
+msize = int(model_size, i4)
+
+call nc_begin_define_mode(ncid)
+
+call nc_add_global_creation_time(ncid)
+
+call nc_add_global_attribute(ncid, "model_source", source )
+call nc_add_global_attribute(ncid, "model_revision", revision )
+call nc_add_global_attribute(ncid, "model_revdate", revdate )
+
+call nc_add_global_attribute(ncid, "model", "Lorenz_96_Tracer_Advection")
+call nc_add_global_attribute(ncid, "model_forcing", forcing )
+call nc_add_global_attribute(ncid, "model_delta_t", delta_t )
+call nc_add_global_attribute(ncid, "source_rate", source_rate)
+call nc_add_global_attribute(ncid, "exponential_sink_folding", e_folding)
+
+call nc_write_location_atts(ncid, msize/3)
+call nc_end_define_mode(ncid)
+call nc_write_location(ncid, state_loc(1:40), msize/3)
```

Comment:
```suggestion
call nc_write_location_atts(ncid, grid_size)
call nc_end_define_mode(ncid)
call nc_write_location(ncid, state_loc(1:40), grid_size)
```

Reply:
i have a question about line 523 - the second argument to write_location should be an array that's the same size as the integer 3rd location.  is 'grid_size' always going to be less than 40?  it seems like it could be:


> call nc_write_location(ncid, state_loc(1:grid_size), grid_size)  
> or 
> call nc_write_location(ncid, state_loc, grid_size)

i don't know what's going on at this point in the code, but the 1:40 seems like a red flag.

Reply:
yup yup yup,  I just saw a straggling /3 so I put this note.
1:40 is dangerous.

I think the model size is a bit weird.  model_size is a namelist option, so you put 120 for the 40 lorenz * 3 variables.  But should you set the 40?   
Also it is not a grid so grid_size probably needs a better name.

Reply:
There should be no hard-coded sizes in the model_mod code. The model is composed of 3 equal size sub state fields, the wind (standard L96), tracer concentration and tracer source. The fact that the length of each field is 1/3 of the total model_size can be used, but there should be no use of any hard-coded field length. I believe the fix for the interpolation was already commited, but it had this same type of problem.